checker/requestcache.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"fmt"
+	"regexp"
+
+	authz "github.com/envoyproxy/data-plane-api/api/auth"
+	"github.com/projectcalico/app-policy/policystore"
+	"github.com/projectcalico/app-policy/proto"
+)
+
+// requestCache contains the CheckRequest and cached copies of computed information about the request
+type requestCache struct {
+	Request *authz.CheckRequest
+	store   *policystore.PolicyStore
+	source  *Peer
+}
+
+type Peer struct {
+	Name      string
+	Namespace string
+	Labels    map[string]string
+}
+
+// SPIFFE_ID_PATTERN is a regular expression to match SPIFFE ID URIs, e.g. spiffe://cluster.local/ns/default/sa/foo
+const SPIFFE_ID_PATTERN = "^spiffe://[^/]+/ns/([^/]+)/sa/([^/]+)$"
+
+var spiffeIdRegExp *regexp.Regexp
+
+func NewRequestCache(store *policystore.PolicyStore, req *authz.CheckRequest) *requestCache {
+	return &requestCache{Request: req, store: store}
+}
+
+// SourceName returns the cached source name. You must call InitSource() before calling this method.
+func (r *requestCache) Source() Peer {
+	if r.source == nil {
+		panic("Called Source() before InitSource()")
+	}
+	return *r.source
+}
+
+// InitSource initializes the source peer. It parses the SPIFFE ID of the source peer, and stores the result.  It
+// accesses labels from the store and merges them with the request (if any).  Idempotent.
+func (r *requestCache) InitSource() error {
```

Comment:
may be rename to just `init`; that way we can add more init's to it and then not have to rename the api.